### PR TITLE
Handle socket errors and better handling of v4

### PR DESCRIPTION
**Pull request by thomasrockhu**

### CODE REVIEW

File: lib/codecov.rb [hunk 1]
```diff
@@ -362,8 +363,8 @@ def upload_to_v4(url, report, query, query_without_token)
       }
     )
     response = retry_request(req, https)
-    if response.code == '400'
-      puts response.body.red
+    if !response&.code || response.code == '400'
```
Comment (Hemaraj-Gajaraj) [inline lib/codecov.rb:366]: @thomasrockhu I am getting syntax error here while I am trying to setup codecov. Whether the format used is valid?

Comment (thomasrockhu) [reply lib/codecov.rb:366]: @Hemaraj-Gajaraj could you link me to the error?

Comment (Hemaraj-Gajaraj) [reply lib/codecov.rb:366]: @thomasrockhu I am trying to setup codecov in my dev environment and faces this issue.
rvm/gems/ruby-2.2.5/gems/codecov-0.2.2/lib/codecov.rb:366: syntax error, unexpected '.' (SyntaxError), 
rvm/gems/ruby-2.2.5/gems/codecov-0.2.2/lib/codecov.rb:367: syntax error, unexpected '.'

Comment (thomasrockhu) [reply lib/codecov.rb:366]: @Hemaraj-Gajaraj, the latest version is `0.2.5`. Also, are you on Ruby 2.2.5? I believe that has been end of lifed 2 years ago

Comment (thomasrockhu) [reply lib/codecov.rb:366]: The syntax of `&.` was introduced in 2.3 http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/

Comment (AlexWayfer) [reply lib/codecov.rb:366]: > The syntax of `&.` was introduced in 2.3 http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/

So, you have to specify Ruby version requirement in the gemspec: https://guides.rubygems.org/specification-reference/#required_ruby_version=

I don't see it in the current `master`: https://github.com/codecov/codecov-ruby/blob/bc8789a/codecov.gemspec

Comment (thomasrockhu) [reply lib/codecov.rb:366]: Thanks @AlexWayfer, I'll make that change

**Pull request merged**